File: haskell/test/TestData.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedStrings #-}

module TestData
  ( PointDistance (..),
    SegmentDistance (..),
    Simplification (..),
    AlgorithmicSimplification (..),
    points,
    pointDistances,
    segmentDistances,
    radialSimplifications,
    douglasPeuckerSimplifications,
    algorithmSimplifications,
  )
where

import Control.Monad (MonadPlus (mzero))
import Data.Aeson
  ( FromJSON (parseJSON),
    Value (Object),
    eitherDecode,
    (.:),
  )
import qualified Data.ByteString.Lazy as B
import GHC.Generics (Generic)
import Impl.Simplify (Point (Point2d))

instance FromJSON Point where
  parseJSON (Object json) =
    Point2d
      <$> json .: "x"
        <*> json .: "y"

data PointDistance = PointDistance
  { left :: Point,
    right :: Point,
    distance :: Double
  }
  deriving (Generic)

data SegmentDistance = SegmentDistance
  { point :: Point,
    start :: Point,
    end :: Point,
    distance :: Double
  }
  deriving (Generic)

data Simplification = Simplification
  { tolerance :: Double,
    result :: [Point]
  }
  deriving (Generic)

data AlgorithmicSimplification = AlgorithmicSimplification
  { tolerance :: Double,
    highQuality :: Bool,
    result :: [Point]
  }
  deriving (Generic)

instance FromJSON PointDistance

instance FromJSON SegmentDistance

instance FromJSON Simplification

instance FromJSON AlgorithmicSimplification

pointsFile :: FilePath
pointsFile = "../test-data/points.json"

distanceFile :: FilePath
distanceFile = "../test-data/point-distances.json"

segmentFile :: FilePath
segmentFile = "../test-data/segment-distances.json"

radialSimplificationFile :: FilePath
radialSimplificationFile = "../test-data/radial-simplifications.json"

douglasPeuckerSimplificationFile :: FilePath
douglasPeuckerSimplificationFile = "../test-data/douglas-peucker-simplifications.json"

algorithmicSimplificationFile :: FilePath
algorithmicSimplificationFile = "../test-data/algorithm-simplifications.json"

getJSON :: FilePath -> IO B.ByteString
getJSON = B.readFile

points :: IO [Point]
points = do
  json <- (eitherDecode <$> getJSON pointsFile) :: IO (Either String [Point])
  case json of
    Left err -> error err
    Right values -> return values

pointDistances :: IO [PointDistance]
pointDistances = do
  json <- (eitherDecode <$> getJSON distanceFile) :: IO (Either String [PointDistance])
  case json of
    Left err -> error err
    Right values -> return values

segmentDistances :: IO [SegmentDistance]
segmentDistances = do
  json <- (eitherDecode <$> getJSON segmentFile) :: IO (Either String [SegmentDistance])
  case json of
    Left err -> error err
    Right values -> return values

radialSimplifications :: IO [Simplification]
radialSimplifications = do
  json <- (eitherDecode <$> getJSON radialSimplificationFile) :: IO (Either String [Simplification])
  case json of
    Left err -> error err
    Right values -> return values

douglasPeuckerSimplifications :: IO [Simplification]
douglasPeuckerSimplifications = do
  json <- (eitherDecode <$> getJSON douglasPeuckerSimplificationFile) :: IO (Either String [Simplification])
  case json of
    Left err -> error err
    Right values -> return values

algorithmSimplifications :: IO [AlgorithmicSimplification]
algorithmSimplifications = do
  json <- (eitherDecode <$> getJSON algorithmicSimplificationFile) :: IO (Either String [AlgorithmicSimplification])
  case json of
    Left err -> error err
    Right values -> return values

File: dart/test/test_data.dart
import 'dart:convert';
import 'dart:io';

import 'package:simplify_dart/simplify.dart';

class _PointDistance {
  final Point left;
  final Point right;
  final num distance;

  _PointDistance.fromJson(Map<String, dynamic> json)
      : distance = json['distance'],
        left = Point2d(json['left']['x'], json['left']['y']),
        right = Point2d(json['right']['x'], json['right']['y']);
}

class _SegmentDistance {
  final Point point;
  final Point start;
  final Point end;
  final num distance;

  _SegmentDistance.fromJson(Map<String, dynamic> json)
      : distance = json['distance'],
        point = Point2d(json['point']['x'], json['point']['y']),
        start = Point2d(json['start']['x'], json['start']['y']),
        end = Point2d(json['end']['x'], json['end']['y']);
}

class _AlgorithmicSimplification {
  final num tolerance;
  final bool highQuality;
  final List<Point> result;

  _AlgorithmicSimplification.fromJson(Map<String, dynamic> json)
      : tolerance = json['tolerance'],
        highQuality =
            json.containsKey('highQuality') ? json['highQuality'] : false,
        result = (json['result'] as List<dynamic>)
            .map((e) => Point2d(e['x'], e['y']))
            .toList();
}

List<_PointDistance> getDistances() {
  final content = File('../test-data/point-distances.json').readAsStringSync();
  final json = jsonDecode(content);
  final result = <_PointDistance>[];

  for (var object in json) {
    result.add(_PointDistance.fromJson(object));
  }

  return result;
}

List<_SegmentDistance> getSegmentDistances() {
  final content =
      File('../test-data/segment-distances.json').readAsStringSync();
  final json = jsonDecode(content);
  final result = <_SegmentDistance>[];

  for (var object in json) {
    result.add(_SegmentDistance.fromJson(object));
  }

  return result;
}

List<Point> getPoints() {
  final content = File('../test-data/points.json').readAsStringSync();
  final json = jsonDecode(content);
  final result = <Point>[];

  for (var object in json) {
    result.add(Point2d(object['x'], object['y']));
  }

  return result;
}

List<_AlgorithmicSimplification> getRadialSimplifications() {
  final content =
      File('../test-data/radial-simplifications.json').readAsStringSync();
  final json = jsonDecode(content);
  final result = <_AlgorithmicSimplification>[];

  for (var object in json) {
    result.add(_AlgorithmicSimplification.fromJson(object));
  }

  return result;
}

List<_AlgorithmicSimplification> getDouglasPeuckerSimplifications() {
  final content = File('../test-data/douglas-peucker-simplifications.json')
      .readAsStringSync();
  final json = jsonDecode(content);
  final result = <_AlgorithmicSimplification>[];

  for (var object in json) {
    result.add(_AlgorithmicSimplification.fromJson(object));
  }

  return result;
}

List<_AlgorithmicSimplification> getAlgorithmicSimplifications() {
  final content =
      File('../test-data/algorithm-simplifications.json').readAsStringSync();
  final json = jsonDecode(content);
  final result = <_AlgorithmicSimplification>[];

  for (var object in json) {
    result.add(_AlgorithmicSimplification.fromJson(object));
  }

  return result;
}
